activerecord/lib/active_record/attribute_methods/write.rb

```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
         end
 
         name = self.class.primary_key if name == "id".freeze && self.class.primary_key
-        @attributes.write_from_user(name, value)
-        value
+        _write_attribute(name, value)
       end
 
       def raw_write_attribute(attr_name, value) # :nodoc:
```

Comment:
Renaming `raw_write_attribute` to `write_attribute_without_type_cast` is probably the simplest thing we could do here to make the interface less ambiguous.